tests/tensorbox/test_tensorbox.py

```diff
@@ -0,0 +1,541 @@
+
+        xT = qml.TensorBox(x)
+        res = xT * y
+        assert np.allclose(res.unbox(), x * y)
+
+        yT = qml.TensorBox(y)
+        res = x * yT
+        assert np.allclose(res.unbox(), x * y)
+
+        res = xT * yT
+        assert np.allclose(res.unbox(), x * y)
+
+    def test_unbox_list(self, tf):
+        """Test unboxing a mixed list works correctly"""
+        x = tf.Variable([[1, 2], [3, 4]])
+        y = tf.Variable([[1, 0], [0, 1]])
+
+        xT = qml.TensorBox(x)
+        res = xT.unbox_list([y, xT, x])
+
+        assert np.all(res == [y, x, x])
+
+    def test_numpy(self, tf):
+        """Test that calling numpy() returns a NumPy array representation
+        of the TensorBox"""
+        x = tf.Variable([[1, 2], [3, 4]])
+        xT = qml.TensorBox(x)
+        assert isinstance(xT.numpy(), np.ndarray)
+        assert not isinstance(xT.numpy(), tf.Variable)
+        assert np.all(xT.numpy() == x.numpy())
+
+    def test_shape(self, tf):
+        """Test that arrays return the right shape"""
+        x = tf.Variable([[[1, 2], [3, 4]]])
+        x = qml.TensorBox(x)
+        res = x.shape
+        assert res == (1, 2, 2)
+
+    def test_expand_dims(self, tf):
+        """Test that dimension expansion works"""
+        from pennylane.tensorbox.tf_box import TensorFlowBox
+
+        x = tf.Variable([1, 2, 3])
+        xT = qml.TensorBox(x)
+
+        res = xT.expand_dims(axis=1)
+        expected = tf.expand_dims(x, 1)
+        assert isinstance(res, TensorFlowBox)
+        assert np.allclose(res.unbox(), expected)
+
+    def test_ones_like(self, tf):
+        """Test that all ones arrays are correctly created"""
+        from pennylane.tensorbox.tf_box import TensorFlowBox
+
+        x = tf.Variable([[1, 2, 3], [4, 5, 6]])
+        xT = qml.TensorBox(x)
+
+        res = xT.ones_like()
+        expected = tf.ones_like(x)
+        assert isinstance(res, TensorFlowBox)
+        assert np.allclose(res.unbox(), expected)
+
+    def test_stack(self, tf):
+        """Test that arrays are correctly stacked together"""
+        x = tf.Variable([[1, 2], [3, 4]])
+        y = tf.Variable([[1, 0], [0, 1]])
+
+        xT = qml.TensorBox(x)
+        res = xT.stack([y, xT, x])
+
+        assert np.allclose(res.unbox(), tf.stack([y, x, x]))
+
+    def test_transpose(self, tf):
+        """Test that the transpose is correct"""
+        x = tf.Variable([[1, 2], [3, 4]])
+        xT = qml.TensorBox(x)
+
+        assert np.allclose(xT.T.unbox(), tf.transpose(x))
+
+    def test_autodifferentiation(self, tf):
+        """Test that autodifferentiation is preserved when writing
+        a cost function that uses TensorBox method chaining"""
+        x = tf.Variable([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]])
+
+        with tf.GradientTape() as tape:
+            y = (qml.TensorBox(x).T ** 2).unbox()[0, 1]
+
+        res = tape.gradient(y, x)
+        expected = tf.constant([[0.0, 0.0, 0.0], [8.0, 0.0, 0.0]])
+        assert np.allclose(res, expected)

```

Comment:
I don't want to be pedantic, but the biggest danger in this class is that the different interfaces implement the same function differently. I don't remember a good example, but had it before that how dims/axes and other args are interpreted is not 100% the same...well, didn't we have that problem with tolerances before?

So, I wonder if it is worth adding a test class that checks whether the numpy results are all the same, using some edge test cases?

Reply:
Do we need a test for chaining?

Reply:
@mariaschuld I've now added a new test suite (`test_functions.py`) that tests the functions using parametrized test data from all interfaces, to make sure they agree on the result!